fili-core/src/main/java/com/yahoo/bard/webservice/table/availability/BaseCompositeAvailability.java

```diff
@@ -0,0 +1,70 @@
+// Copyright 2017 Yahoo Inc.
+// Licensed under the terms of the Apache license. Please see LICENSE.md file distributed with this work for terms.
+package com.yahoo.bard.webservice.table.availability;
+
+import com.yahoo.bard.webservice.data.config.names.TableName;
+import com.yahoo.bard.webservice.util.SimplifiedIntervalList;
+import com.yahoo.bard.webservice.util.StreamUtils;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * A class implementing common capabilities for availabilities backed by a collection of other availabilities.
+ */
+public abstract class BaseCompositeAvailability implements Availability {
+
+    private final Set<Availability> sourceAvailabilities;
+    private final Set<TableName> dataSourcesNames;
+
+    /**
+     * Constructor.
+     *
+     * @param availabilityStream  A potentially ordered stream of availabilities which supply this composite view
+     */
+    protected BaseCompositeAvailability(Stream<Availability> availabilityStream) {
+        sourceAvailabilities = StreamUtils.toUnmodifiableSet(availabilityStream);
+        dataSourcesNames = StreamUtils.toUnmodifiableSet(
+                sourceAvailabilities.stream()
+                        .map(Availability::getDataSourceNames).flatMap(Set::stream)
+        );
+    }
+
+    /**
+     * Return a stream of all the availabilities which this availability composites from.
+     *
+     * @return A stream of availabilities
+     */
+    protected Stream<Availability> getAllSourceAvailabilities() {
```

Comment:
There's no "store" going on here, simply passing of information. If we're expecting the caller to convert any collection we give them into a stream for processing anyways, why not give them the stream right away, rather than collecting into a collection, and then streaming out again after that?